File: golang/internal/application/service/portfolio_service.go
package service

import (
	"fmt"

	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

func floor2(x decimal.Decimal) decimal.Decimal {
	return x.Truncate(2)
}

func floor0(x decimal.Decimal) decimal.Decimal {
	return x.Truncate(0)
}

func priceOf(prices map[domain.StockSymbol]decimal.Decimal, symbol domain.StockSymbol) (decimal.Decimal, error) {
	p, ok := prices[symbol]
	if !ok {
		return decimal.Zero, fmt.Errorf("price not found for symbol: %s", symbol)
	}
	return p, nil
}

func AllocateNew(amount decimal.Decimal, portfolio domain.Portfolio, prices map[domain.StockSymbol]decimal.Decimal) (domain.Account, error) {
	cashFromRate := floor0(amount.Mul(domain.CashRate))
	investable := amount.Sub(cashFromRate)

	stocks := make([]domain.Stock, 0, len(portfolio.Items))
	for _, item := range portfolio.Items {
		price, err := priceOf(prices, item.Symbol)
		if err != nil {
			return domain.Account{}, err
		}
		qty := floor2(investable.Mul(item.Rate).Div(price))
		stocks = append(stocks, domain.Stock{Symbol: item.Symbol, Qty: qty})
	}

	usedForStocks := decimal.Zero
	for _, s := range stocks {
		price, err := priceOf(prices, s.Symbol)
		if err != nil {
			return domain.Account{}, err
		}
		usedForStocks = usedForStocks.Add(s.Qty.Mul(price))
	}
	residual := investable.Sub(usedForStocks)

	return domain.Account{Cash: cashFromRate.Add(residual), Stocks: stocks}, nil
}

func AllocateAdditional(account domain.Account, amount decimal.Decimal, portfolio domain.Portfolio, prices map[domain.StockSymbol]decimal.Decimal) (domain.Account, error) {
	totalAfterVal, err := TotalValuation(account, prices)
	if err != nil {
		return domain.Account{}, err
	}
	totalAfter := totalAfterVal.Add(amount)
	targetCash := floor0(totalAfter.Mul(domain.CashRate))
	investable := totalAfter.Sub(targetCash)

	currentQty := make(map[domain.StockSymbol]decimal.Decimal)
	for _, s := range account.Stocks {
		currentQty[s.Symbol] = s.Qty
	}

	portfolioSymbols := make(map[domain.StockSymbol]struct{})
	for _, item := range portfolio.Items {
		portfolioSymbols[item.Symbol] = struct{}{}
	}

	newPortfolioStocks := make([]domain.Stock, 0, len(portfolio.Items))
	for _, item := range portfolio.Items {
		price, err := priceOf(prices, item.Symbol)
		if err != nil {
			return domain.Account{}, err
		}
		targetQty := floor2(investable.Mul(item.Rate).Div(price))
		current := currentQty[item.Symbol]
		finalQty := targetQty
		if current.GreaterThan(targetQty) {
			finalQty = current
		}
		newPortfolioStocks = append(newPortfolioStocks, domain.Stock{Symbol: item.Symbol, Qty: finalQty})
	}

	preservedStocks := make([]domain.Stock, 0)
	for _, s := range account.Stocks {
		if _, inPortfolio := portfolioSymbols[s.Symbol]; !inPortfolio {
			preservedStocks = append(preservedStocks, s)
		}
	}

	allStocks := append(newPortfolioStocks, preservedStocks...)

	finalValuation := decimal.Zero
	for _, s := range allStocks {
		price, err := priceOf(prices, s.Symbol)
		if err != nil {
			return domain.Account{}, err
		}
		finalValuation = finalValuation.Add(s.Qty.Mul(price))
	}
	finalCash := totalAfter.Sub(finalValuation)

	return domain.Account{Cash: finalCash, Stocks: allStocks}, nil
}

func Rebalance(account domain.Account, portfolio domain.Portfolio, prices map[domain.StockSymbol]decimal.Decimal) (domain.Account, error) {
	// XXX this implementation might not be correct
	investable, err := TotalValuation(account, prices)
	if err != nil {
		return domain.Account{}, err
	}

	newStocks := make([]domain.Stock, 0, len(portfolio.Items))
	for _, item := range portfolio.Items {
		price, err := priceOf(prices, item.Symbol)
		if err != nil {
			return domain.Account{}, err
		}
		qty := floor2(investable.Mul(item.Rate).Div(price))
		newStocks = append(newStocks, domain.Stock{Symbol: item.Symbol, Qty: qty})
	}

	finalValuation := decimal.Zero
	for _, s := range newStocks {
		price, err := priceOf(prices, s.Symbol)
		if err != nil {
			return domain.Account{}, err
		}
		finalValuation = finalValuation.Add(s.Qty.Mul(price))
	}
	finalCash := investable.Sub(finalValuation)

	return domain.Account{Cash: finalCash, Stocks: newStocks}, nil
}

File: php/src/CodingInterview/Application/Service/PortfolioService.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Service;

use Folio\CodingInterview\Domain\Account;
use Folio\CodingInterview\Domain\AppConstants;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\Stock;
use Folio\CodingInterview\Domain\StockSymbol;
use Folio\CodingInterview\Domain\Portfolio;

final class PortfolioService
{
    private static function floor2(BigDecimal $x): BigDecimal
    {
        return $x->setScaleDown(2);
    }

    private static function floor0(BigDecimal $x): BigDecimal
    {
        return $x->setScaleDown(0);
    }

    /** @param array<string,BigDecimal> $prices */
    private static function priceOf(array $prices, StockSymbol $symbol): BigDecimal
    {
        if (!array_key_exists($symbol->value, $prices)) {
            throw new \LogicException("missing price for {$symbol->value}");
        }
        return $prices[$symbol->value];
    }

    /**
     * Allocate a brand-new account given a contribution amount.
     * @param array<string,BigDecimal> $prices
     */
    public static function allocateNew(BigDecimal $amount, Portfolio $portfolio, array $prices): Account
    {
        $cashFromRate = self::floor0($amount->mul(AppConstants::cashRate()));
        $investable = $amount->sub($cashFromRate);
        $stocks = [];
        foreach ($portfolio->items as $item) {
            $price = self::priceOf($prices, $item->symbol);
            $qty = self::floor2($investable->mul($item->rate)->div($price));
            $stocks[] = new Stock($item->symbol, $qty);
        }
        $usedForStocks = BigDecimal::zero();
        foreach ($stocks as $e) {
            $usedForStocks = $usedForStocks->add($e->qty->mul(self::priceOf($prices, $e->symbol)));
        }
        $residual = $investable->sub($usedForStocks);
        return new Account($cashFromRate->add($residual), $stocks);
    }

    /**
     * Additional contribution: only buy (no sell). Residual is kept in cash.
     * @param array<string,BigDecimal> $prices
     */
    public static function allocateAdditional(
        Account $account,
        BigDecimal $amount,
        Portfolio $portfolio,
        array $prices,
    ): Account {
        $totalAfter = AssetService::totalValuation($account, $prices)->add($amount);
        $targetCash = self::floor0($totalAfter->mul(AppConstants::cashRate()));
        $investable = $totalAfter->sub($targetCash);

        $currentQty = [];
        foreach ($account->stocks as $e) {
            $currentQty[$e->symbol->value] = $e->qty;
        }

        $portfolioSymbols = [];
        foreach ($portfolio->items as $i) {
            $portfolioSymbols[$i->symbol->value] = true;
        }
        $newPortfolioStocks = [];
        foreach ($portfolio->items as $item) {
            $price = self::priceOf($prices, $item->symbol);
            $targetQty = self::floor2($investable->mul($item->rate)->div($price));
            $current = $currentQty[$item->symbol->value] ?? BigDecimal::zero();
            $finalQty = $targetQty->gt($current) ? $targetQty : $current;
            $newPortfolioStocks[] = new Stock($item->symbol, $finalQty);
        }
        $preservedStocks = [];
        foreach ($account->stocks as $e) {
            if (!isset($portfolioSymbols[$e->symbol->value])) {
                $preservedStocks[] = $e;
            }
        }
        $allStocks = array_merge($newPortfolioStocks, $preservedStocks);

        $finalValuation = BigDecimal::zero();
        foreach ($allStocks as $e) {
            $finalValuation = $finalValuation->add($e->qty->mul(self::priceOf($prices, $e->symbol)));
        }
        $finalCash = $totalAfter->sub($finalValuation);
        return new Account($finalCash, $allStocks);
    }

    /**
     * Rebalance: re-allocate qty per portfolio target (buy and sell).
     * @param array<string,BigDecimal> $prices
     */
    public static function rebalance(Account $account, Portfolio $portfolio, array $prices): Account
    {
        // XXX this implementation might not be correct
        $investable = AssetService::totalValuation($account, $prices);
        $newStocks = [];
        foreach ($portfolio->items as $item) {
            $price = self::priceOf($prices, $item->symbol);
            $qty = self::floor2($investable->mul($item->rate)->div($price));
            $newStocks[] = new Stock($item->symbol, $qty);
        }
        $finalValuation = BigDecimal::zero();
        foreach ($newStocks as $e) {
            $finalValuation = $finalValuation->add($e->qty->mul(self::priceOf($prices, $e->symbol)));
        }
        $finalCash = $investable->sub($finalValuation);
        return new Account($finalCash, $newStocks);
    }
}
